project3/participant/tasks/multicast_receiver_task.cpp

```diff
@@ -0,0 +1,46 @@
+#include "multicast_receiver_task.h"
+
+#include <loguru.hpp>
+
+namespace participant_tasks {
+MulticastReceiver::MulticastReceiver(
+    const std::shared_ptr<ConsoleTask>& console_task, std::string log_location,
+    int port)
+    : port_(port), log_location_(log_location), console_task_(console_task) {}
+
+thread_pool::Task::Status MulticastReceiver::SetUp() {
+  struct sockaddr_in address = participant_util::MakeAddress(port_);
+  messenger_fd_ = participant_util::SetUpListenerSocket(address);
+  loguru::add_file(log_location_.c_str(), loguru::FileMode::Truncate,
+                   loguru::Verbosity_0);
+  return thread_pool::Task::Status::RUNNING;
+}
+
+thread_pool::Task::Status MulticastReceiver::RunAtomic() {
+  parser_.ResetParser();
+  while (!parser_.HasCompleteMessage()) {
+    incoming_msg_buf_.resize(kBufferSize);
+    const auto bytes_read = participant_util::ReceiveForever(
+        messenger_fd_, incoming_msg_buf_.data(), kBufferSize, 0);
+    if (bytes_read <= 0) {
+      break;
+    }
+    incoming_msg_buf_.resize(bytes_read);
+    parser_.AddNewData(incoming_msg_buf_);
+  }
+  // empty buffer
+  incoming_msg_buf_.clear();
+  if (parser_.HasCompleteMessage()) {
```

Comment:
Isn't this condition always going to be true by the time the above loop exits? From a design perspective, it's probably better to not reset the parser on every call to this function, and just have it run one iteration of the receiving loop on each `RunAtomic` call. It's simpler code and it fits more with the `Task` design philosophy.